src/vm/functions/database.rs

```diff
@@ -8,34 +8,99 @@ use vm::representations::{SymbolicExpression, SymbolicExpressionType};
-    let mut rest_args = rest_args?;
-    let rest_args: Vec<_> = rest_args.drain(..).map(|x| { SymbolicExpression::atom_value(x) }).collect();
+    let (contract_identifier, type_returns_constraint) = match &args[0].expr {
+        SymbolicExpressionType::LiteralValue(Value::Principal(PrincipalData::Contract(ref contract_identifier))) => {
+            // Static dispatch
+            (contract_identifier, None)
+        },
+        SymbolicExpressionType::Atom(contract_ref) => {
+            // Dynamic dispatch
+            match context.callable_contracts.get(contract_ref) {
+                Some((ref contract_identifier, trait_identifier)) => {
+                    // Ensure that contract-call is used for inter-contract calls only 
+                    if *contract_identifier == env.contract_context.contract_identifier {
+                        return Err(CheckErrors::CircularReference(vec![contract_identifier.name.to_string()]).into());
+                    }
+                    
+                    let contract_to_check = env.global_context.database.get_contract(contract_identifier)
+                        .map_err(|_e| CheckErrors::NoSuchContract(contract_identifier.to_string()))?;
+                    let contract_context_to_check = contract_to_check.contract_context;
+
+                    // Attempt to short circuit the dynamic dispatch checks:
+                    // If the contract is explicitely implementing the trait with `impl-trait`,
+                    // then we can simply rely on the analysis performed at publish time.
+                    if contract_context_to_check.is_explicitly_implementing_trait(&trait_identifier) {
+                        (contract_identifier, None)
+                    } else {
+                        let trait_name = trait_identifier.name.to_string();
+
+                        // Retrieve, from the trait definition, the expected method signature
+                        let contract_defining_trait = env.global_context.database.get_contract(&trait_identifier.contract_identifier)
+                            .map_err(|_e| CheckErrors::NoSuchContract(trait_identifier.contract_identifier.to_string()))?;
+                        let contract_context_defining_trait = contract_defining_trait.contract_context;
+
+                        // Retrieve the function that will be invoked
+                        let function_to_check = contract_context_to_check.lookup_function(function_name)
+                            .ok_or(CheckErrors::BadTraitImplementation(trait_name.clone(), function_name.to_string()))?;
+                        
+                        // Check read/write compatibilty: if we are in a readonly context, then the next function should be readonly.
+                        // Otherwise the method should be public
+                        match (env.global_context.is_read_only(), &function_to_check.define_type) {
```

Comment:
We should always fail out of the read-only context if doing a dynamic contract-call.

We can support read-only trait function calls through better and more explicit mechanisms than a dynamic check here.

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1246/commits/3f498088a35770771a5e74b5a2cf230a87d11950